Source/FieldSolver/ImplicitSolvers/WarpXImplicitOps.cpp

```diff
@@ -0,0 +1,333 @@
+}
+
+void
+WarpX::UpdateMagneticFieldAndApplyBCs( const amrex::Vector<std::array< std::unique_ptr<amrex::MultiFab>, 3 > >&  a_Bn,
+                            amrex::Real                                                         a_thetadt )
+{
+    amrex::MultiFab::Copy(*Bfield_fp[0][0], *a_Bn[0][0], 0, 0, ncomps, a_Bn[0][0]->nGrowVect());
+    amrex::MultiFab::Copy(*Bfield_fp[0][1], *a_Bn[0][1], 0, 0, ncomps, a_Bn[0][1]->nGrowVect());
+    amrex::MultiFab::Copy(*Bfield_fp[0][2], *a_Bn[0][2], 0, 0, ncomps, a_Bn[0][2]->nGrowVect());
+    EvolveB(a_thetadt, DtType::Full);
+    ApplyMagneticFieldBCs();
+}
+
+void
+WarpX::FinishMagneticFieldAndApplyBCs( const amrex::Vector<std::array< std::unique_ptr<amrex::MultiFab>, 3 > >&  a_Bn,
+                            amrex::Real                                                         a_theta )
+{
+    FinishImplicitField(Bfield_fp, a_Bn, a_theta);
+    ApplyMagneticFieldBCs();
+}
+
+void
+WarpX::ApplyMagneticFieldBCs()
+{
+    FillBoundaryB(guard_cells.ng_alloc_EB, WarpX::sync_nodal_points);
+    ApplyBfieldBoundary(0, PatchType::fine, DtType::Full);
+}
+
+void
+WarpX::SaveParticlesAtImplicitStepStart ( )
+{
+    const int lev = 0;
+
+    for (auto const& pc : *mypc) {
+
+#ifdef AMREX_USE_OMP
+#pragma omp parallel
+#endif
+    {
+
+    auto particle_comps = pc->getParticleComps();
+
+    for (WarpXParIter pti(*pc, lev); pti.isValid(); ++pti) {
+
+        const auto getPosition = GetParticlePosition(pti);
+
+        auto& attribs = pti.GetAttribs();
+        amrex::ParticleReal* const AMREX_RESTRICT ux = attribs[PIdx::ux].dataPtr();
+        amrex::ParticleReal* const AMREX_RESTRICT uy = attribs[PIdx::uy].dataPtr();
+        amrex::ParticleReal* const AMREX_RESTRICT uz = attribs[PIdx::uz].dataPtr();
+
+#if (AMREX_SPACEDIM >= 2)
+        amrex::ParticleReal* x_n = pti.GetAttribs(particle_comps["x_n"]).dataPtr();
+#endif
+#if defined(WARPX_DIM_3D) || defined(WARPX_DIM_RZ)
+        amrex::ParticleReal* y_n = pti.GetAttribs(particle_comps["y_n"]).dataPtr();
+#endif
+        amrex::ParticleReal* z_n = pti.GetAttribs(particle_comps["z_n"]).dataPtr();
+        amrex::ParticleReal* ux_n = pti.GetAttribs(particle_comps["ux_n"]).dataPtr();
+        amrex::ParticleReal* uy_n = pti.GetAttribs(particle_comps["uy_n"]).dataPtr();
+        amrex::ParticleReal* uz_n = pti.GetAttribs(particle_comps["uz_n"]).dataPtr();
+
+        const long np = pti.numParticles();
+
+        amrex::ParallelFor( np, [=] AMREX_GPU_DEVICE (long ip)
+        {
+            amrex::ParticleReal xp, yp, zp;
+            getPosition(ip, xp, yp, zp);
+
+#if (AMREX_SPACEDIM >= 2)
+            x_n[ip] = xp;
+#endif
+#if defined(WARPX_DIM_3D) || defined(WARPX_DIM_RZ)
+            y_n[ip] = yp;
+#endif
+            z_n[ip] = zp;
+
+            ux_n[ip] = ux[ip];
+            uy_n[ip] = uy[ip];
+            uz_n[ip] = uz[ip];
+
+        });
+
+    }
+    }
+
+    }
+
+}
+
+void
+WarpX::FinishImplicitParticleUpdate ()
+{
+    using namespace amrex::literals;
```

Comment:
Could you add a comment here, explaining what this function does
(i.e. at the end of a nonlinear solve, we get the particle momentum and position at n+1/2. We then need to compute the position and momentum at `n+1`, which is what this function does.)